File: Experience vs Salary hike.py
# Importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

# Input Variable = Years of Experience
# Output Varibale = Salary Hike

Sal_hike = pd.read_csv("C:/Users/SHAJIUDDIN MOHAMMED/Desktop/Data sets/Salary_Data.csv")
Sal_hike.columns = "Experience", "Salary" 
Sal_hike.columns

# Exploratory data analysis:
Sal_hike.describe()

# Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 

plt.bar(height = Sal_hike.Salary, x = np.arange(1, 31, 1))
plt.hist(Sal_hike.Salary) #histogram
plt.boxplot(Sal_hike.Salary) #boxplot

plt.bar(height = Sal_hike.Experience, x = np.arange(1, 31, 1))
plt.hist(Sal_hike.Experience) #histogram
plt.boxplot(Sal_hike.Experience) #boxplot

# Scatter plot
plt.scatter(x = Sal_hike['Experience'], y = Sal_hike['Salary'], color = 'red') 

# Correlation
np.corrcoef(Sal_hike.Experience, Sal_hike.Salary) 

# Covariance
cov_output = np.cov(Sal_hike.Experience, Sal_hike.Salary)[0, 1]
cov_output

# Import library
import statsmodels.formula.api as smf

# Simple Linear Regression
model = smf.ols('Salary ~ Experience', data = Sal_hike).fit()
model.summary()

pred1 = model.predict(pd.DataFrame(Sal_hike['Experience']))

# Regression Line
plt.scatter(Sal_hike.Experience, Sal_hike.Salary)
plt.plot(Sal_hike.Experience, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res1 = Sal_hike.Salary - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1

# Model building on Transformed Data(Log Transformation)
# x = log(Experience); y = Salary

plt.scatter(x = np.log(Sal_hike['Experience']), y = Sal_hike['Salary'], color = 'brown')
np.corrcoef(np.log(Sal_hike.Experience), Sal_hike.Salary) 

model2 = smf.ols('Salary ~ np.log(Experience)', data = Sal_hike).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(Sal_hike['Experience']))

# Regression Line
plt.scatter(np.log(Sal_hike.Experience), Sal_hike.Salary)
plt.plot(np.log(Sal_hike.Experience), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res2 = Sal_hike.Salary - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2


# Exponential transformation
# x = Experience; y = log(Salary)

plt.scatter(x = Sal_hike['Experience'], y = np.log(Sal_hike['Salary']), color = 'orange')
np.corrcoef(Sal_hike.Experience, np.log(Sal_hike.Salary))

model3 = smf.ols('np.log(Salary) ~ Experience', data = Sal_hike).fit()
model3.summary()

pred3 = model3.predict(pd.DataFrame(Sal_hike['Experience']))
pred3_at = np.exp(pred3)
pred3_at

# Regression Line
plt.scatter(Sal_hike.Experience, np.log(Sal_hike.Salary))
plt.plot(Sal_hike.Experience, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res3 = Sal_hike.Salary - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3


# Polynomial transformation
# x = Experience; x^2 = Experience*Experience; y = log(Salary)

model4 = smf.ols('np.log(Salary) ~ Experience + I(Experience*Experience)', data = Sal_hike).fit()
model4.summary()

pred4 = model4.predict(pd.DataFrame(Sal_hike))
pred4_at = np.exp(pred4)
pred4_at

# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = Sal_hike.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)


plt.scatter(Sal_hike.Experience, np.log(Sal_hike.Salary))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()


# Error calculation
res4 = Sal_hike.Salary - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4


# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse


# Analysis - SLR is the best fitted model with Low Rmse

from sklearn.model_selection import train_test_split

train, test = train_test_split(Sal_hike, test_size = 0.2)

finalmodel = smf.ols('Salary ~ Experience', data = train).fit()
finalmodel.summary()


# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
test_pred

# Model Evaluation on Test data
test_res = test.Salary - test_pred
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse


# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
train_pred

# Model Evaluation on train data

train_res = train.Salary -  train_pred
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

File: Experience vs Salary hike.R
# Input Variable = Years of Experience
# Output Varibale = Salary Hike

#Importing Dataset

Sal_hike <- read.csv(file.choose(), header = T)
colnames(Sal_hike) <- c("Experience", "Salary") 
View(Sal_hike)
attach(Sal_hike)

# Exploratory data analysis
summary(Sal_hike)

install.packages("Hmisc")
library(Hmisc)
describe(Sal_hike) # To have a quick glance of a data

install.packages("lattice")
library("lattice")

# Graphical exploration
dotplot(Sal_hike$Experience, main = "Dot Plot of Experience")
dotplot(Sal_hike$Salary, main = "Dot Plot of Salary")

#Boxplot Representation

boxplot(Sal_hike$Experience, col = "dodgerblue4",main = "Experience")
boxplot(Sal_hike$Salary, col = "red", horizontal = T, main = "Salary")

#Histogram Representation

hist(Sal_hike$Experience,main = "Experience")
hist(Sal_hike$Salary, main = "Salary")

# Normal QQ plot
qqnorm(Sal_hike$Experience,main = "Experience")
qqline(Sal_hike$Experience,main = "Experience")

qqnorm(Sal_hike$Salary, main = "Salary")
qqline(Sal_hike$Salary, main = "Salary")


# Bivariate analysis
# Scatter plot
plot(Sal_hike$Experience, Sal_hike$Salary, main = "Scatter Plot", col = "Dodgerblue4", 
     col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Experience", 
     ylab = "Salary", pch = 20)  # plot(x,y)

# Correlation Coefficient
cor(Experience,Salary)

# Linear Regression model
reg <- lm(Salary ~ Experience, data = Sal_hike ) # Y ~ X
summary(reg)

confint(reg, level = 0.95)
pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)

View(pred)

# ggplot for adding Regression line for data
library(ggplot2)

ggplot(data = Sal_hike, aes(Experience, Salary) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ x)

# Evaluation the model for fitness 
cor(pred$fit, Sal_hike$Salary)

reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse

# Transformation Techniques(transform the variables to check whether the predicted values are better)
# input = log(x); output = y

plot(log(Experience), Salary)
cor(log(Experience), Salary)

reg_log <- lm(Salary ~ log(Experience), data = Sal_hike)
summary(reg_log)

confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")

pred <- as.data.frame(pred)
cor(pred$fit, Sal_hike$Salary)

rmse <- sqrt(mean(reg_log$residuals^2))
rmse

# Regression line for data
ggplot(data = Sal_hike, aes(log(Experience), Salary) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ log(x))

# Log transformation(y)
# input = x; output = log(y)

plot(Experience, log(Salary))
cor(Experience, log(Salary))

reg_log1 <- lm(log(Salary) ~ Experience ,data= Sal_hike)
summary(reg_log1)

predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))

pred <- exp(predlog)  # Exponential function for antilog
pred <- as.data.frame(pred)
cor(pred$fit, Sal_hike$Salary)

res_log1 = Salary - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse

# Regression line for data
ggplot(data = Sal_hike, aes(Experience, log(Salary)) ) +
  geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)

# Polynomial models(Non Linear)
# input = x & x^2 (2-degree) and output = log(y)

reg2 <- lm(log(Salary) ~ Experience + I(Experience*Experience), data = Sal_hike)
summary(reg2)

predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)

pred <- as.data.frame(pred)
cor(pred$fit, Sal_hike$Salary)

res2 = Salary - pred$fit
rmse <- sqrt(mean(res2^2))
rmse

# Regression line for data
ggplot(data = Sal_hike, aes(Experience, log(Salary)) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))

# Data Partition(Randon Sampling)

n <- nrow(Sal_hike)
n1 <- n * 0.8
n2 <- n - n1

train_ind <- sample(1:n, n1)
train <- Sal_hike[train_ind, ]
test <-  Sal_hike[-train_ind, ]

plot(train$Experience, log(train$Salary))
plot(test$Experience, log(test$Salary))

model <- lm(log(Salary) ~ Experience + I(Experience*Experience), data = train)
summary(model)

confint(model,level=0.95)

log_res <- predict(model,interval = "confidence", newdata = test)

predict_original <- exp(log_res) # converting log values to original values
predict_original <- as.data.frame(predict_original)
test_error <- test$Salary - predict_original$fit # calculate error/residual
test_error

test_rmse <- sqrt(mean(test_error^2))
test_rmse

log_res_train <- predict(model,interval = "confidence", newdata = train)

predict_original_train <- exp(log_res_train) # converting log values to original values
predict_original_train <- as.data.frame(predict_original_train)
train_error <- train$Salary - predict_original_train$fit # calculate error/residual
train_error

train_rmse <- sqrt(mean(train_error^2))
train_rmse
